### PR TITLE
boost: conflict with GCC on macOS

### PR DESCRIPTION
**Pull request by albestro**
See the comments for a more detailed explanation of the conflict.

https://github.com/STEllAR-GROUP/hpx/issues/5442#issuecomment-878889166
https://github.com/STEllAR-GROUP/hpx/issues/5442#issuecomment-878913339

@msimberg is it what you found out, right?

### CODE REVIEW

**Comment by albestro** [comment]
Is it worth restricting it to just boost when the component `boost.system` is built? @hainest @msimberg

**Comment by msimberg** [comment]
> See the comments for a more detailed explanation of the conflict.
> 
> [STEllAR-GROUP/hpx#5442 (comment)](https://github.com/STEllAR-GROUP/hpx/issues/5442#issuecomment-878889166)
> [STEllAR-GROUP/hpx#5442 (comment)](https://github.com/STEllAR-GROUP/hpx/issues/5442#issuecomment-878913339)
> 
> @msimberg is it what you found out, right?

Yes, that was my understanding of it. Might be good to get @eschnett's confirmation, and perhaps wait for Boost 1.77 to check that it's all good.

**Comment by eschnett** [comment]
This change describes the conflict well. I don't know about Boost 1.77 (haven't checked the beta release).

**Comment by haampie** [comment]
Did you see https://github.com/spack/spack/pull/24895? It looks similar

File: var/spack/repos/builtin/packages/boost/package.py [hunk 1]
```diff
@@ -198,6 +198,14 @@ def libs(self):
     # Container's Extended Allocators were not added until 1.56.0
     conflicts('+container', when='@:1.55.99')
 
+    # Boost.System till 1.76 (included) was reliying on mutex, which was not
+    # detected correctly on Darwin platform when using GCC
+    #
+    # More details here:
+    # https://github.com/STEllAR-GROUP/hpx/issues/5442#issuecomment-878889166
+    # https://github.com/STEllAR-GROUP/hpx/issues/5442#issuecomment-878913339
+    conflicts('%gcc', when='@:1.76 platform=darwin')
```
Comment (haampie) [inline var/spack/repos/builtin/packages/boost/package.py]: Can you install `boost~system` with gcc on macos?
```suggestion
    conflicts('%gcc', when='@:1.76 +system platform=darwin')
```

Comment (eschnett) [reply var/spack/repos/builtin/packages/boost/package.py]: Yes, installing `boost ~system %gcc` works on macOS.

**Comment by eschnett** [comment]
@haampie I ran across two separate problems when building Boost with GCC on macOS. They are not related.

**Review by haampie — APPROVED** [review]

**Pull request merged**